Sources/NIOCore/VsockAddress.swift

```diff
@@ -0,0 +1,221 @@
+    public struct ContextID: RawRepresentable, ExpressibleByIntegerLiteral, Hashable, Sendable {
+        public var rawValue: UInt32
+
+        public init(rawValue: UInt32) {
+            self.rawValue = rawValue
+        }
+
+        public init(integerLiteral value: UInt32) {
+            self.init(rawValue: value)
+        }
+
+        /// Wildcard, matches any address.
+        ///
+        /// On all platforms, using this value with `bind(2)` means "any address".
+        ///
+        /// On Darwin platforms, the man page states this can be used with `connect(2)` to mean "this host".
+        public static let any: Self = Self(rawValue: VMADDR_CID_ANY)
+
+        /// The address of the hypervisor.
+        public static let hypervisor: Self = Self(rawValue: UInt32(VMADDR_CID_HYPERVISOR))
+
+        /// The address of the host.
+        public static let host: Self = Self(rawValue: UInt32(VMADDR_CID_HOST))
+
+#if os(Linux)
+        /// The address for local communication (loopback).
+        ///
+        /// This directs packets to the same host that generated them.  This is useful for testing
+        /// applications on a single host and for debugging.
+        ///
+        /// The local context ID obtained with ``getLocalContextID(_:)`` can be used for the same
+        /// purpose, but it is preferable to use ``local``.
+        public static let local: Self = Self(rawValue: UInt32(VMADDR_CID_LOCAL))
+#endif
+
+        /// Get the context ID of the local machine.
+        ///
+        /// On Linux, consider using ``local`` when binding instead of this function.
+        public static func getLocalContextID(_ socket: NIOBSDSocket.Handle) -> Self {
+            Self(rawValue: get_local_vsock_cid(socket))
+        }
+    }
+
+    /// A vsock port number.
+    ///
+    /// The vsock port number differentiates between multiple services running on a single machine.
+    public struct Port: RawRepresentable, ExpressibleByIntegerLiteral, Hashable, Sendable {
+        public var rawValue: UInt32
+
+        public init(rawValue: UInt32) {
+            self.rawValue = rawValue
+        }
+
+        public init(integerLiteral value: UInt32) {
+            self.init(rawValue: value)
+        }
+
+        /// Used to bind to any port number.
+        public static let any: Self = Self(rawValue: VMADDR_PORT_ANY)
+    }
+
+    /// Creates a new vsock address.
+    ///
+    /// - parameters:
+    ///   - cid: the context ID.
+    ///   - port: the target port.
+    /// - returns: the `SocketAddress` for the given context ID and port combination.
+    public init(cid: ContextID, port: Port) {
+        self.init(cid: cid.rawValue, port: port.rawValue)
+    }
+}
+
+extension VsockAddress: CustomStringConvertible {
+    public var description: String {
+        "[VSOCK]\(self.cid):\(self.port)"
+    }
+}
+
+/// Element-wise Equatable conformance using only the struct fields defined in the man page (excluding lengths).
+extension VsockAddress: Equatable {
```

Comment:
If we change the storage representation to only store context ID and port then we can use the synthesised Equatable and Hashable implementations.